core/mm3Types.js

```diff
@@ -65,10 +163,61 @@ CLASS({
       help: 'Regex pattern for property.'
     },
     {
+      name: 'minChars',
+      label: 'Minimum characters',
+      help: 'The minimum number of characters required.'
+    },
+    {
+      name: 'maxChars',
+      label: 'Maximum characters',
+      help: 'The maximum number of characters allowed.'
+    },
+    {
       name: 'prototag',
       label: 'Protobuf tag',
       required: false,
       help: 'The protobuf tag number for this field.'
+    },
+    {
+      name: 'validate',
+      lazyFactory: function() {
+        var prop = this; // this == the property
+        var ret = constantFn('');
+
+        var min = prop.minChars;
+        if ( typeof min !== 'string' || min ) { // ignore '' (falsey string)
```

Comment:
The min/max could come from a defaultValue set in the property, or be
user-set on the property instance, so a "null" value is easiest. Yes it
would be better to have validation on the min/max that matches the owner
Property's validation.

On Thu, Dec 10, 2015 at 10:46 AM adamvy <notifications@github.com> wrote:

> In core/mm3Types.js
> <https://github.com/foam-framework/foam/pull/491#discussion_r47242042>:
>
> >        name: 'prototag',
> >        label: 'Protobuf tag',
> >        required: false,
> >        help: 'The protobuf tag number for this field.'
> > +    },
> > +    {
> > +      name: 'validate',
> > +      lazyFactory: function() {
> > +        var prop = this; // this == the property
> > +        var ret = constantFn('');
> > +
> > +        var min = prop.minChars;
> > +        if ( typeof min !== 'string' || min ) { // ignore '' (falsey string)
>
> I'm not suggesting -Infinity +Infinity.
>
> If this wasn't during bootstrap, we could make this a real IntProperty and
> check hasOwnProperty to see if the value is set or not, then only build the
> validate function if its not set. We can't make it a real IntProperty but
> we can still check hasOwnProperty, or see if its an empty string (the
> default for all properties) to see if its set.
>
> Any time we set this, it should be set to an int. The views should be
> setting it to an int, and it could optionally have an adapt that calls
> parseInt.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/foam-framework/foam/pull/491/files#r47242042>.
>